tap-version-14-specification.md

```diff
@@ -0,0 +1,659 @@
+---
+layout: default
+title: TAP 14 specification
+---
+
+# STATUS: DRAFT
+
+This is a draft document pending further editing and ratification by no
+fewer than 3 widely used TAP implementations in 3 different language
+communities.
+
+This section will likely be removed prior to ratification.
+
+## GOALS OF THIS SPECIFICATION
+
+* Document the observed behavior of widely used TAP implementations.
+* Add no features that are not already in wide usage across multiple
+  implementations.
+* Explicitly allow what is already allowed, deny what is already denied.
+* Provide updated and clearer guidance for new TAP implementations.
+
+# NAME
+
+TAP14 - The Test Anything Protocol v14
+
+## SYNOPSIS
+
+TAP, the Test Anything Protocol, is a simple text-based interface between
+testing modules a test harness. TAP started life as part of the test
+harness for Perl but now has implementations in C/C++, Python, PHP, Perl,
+JavaScript, and probably others by the time you read this.  This document
+describes version 14 of TAP. Go to TAP to read about previous versions.
+
+The key words _must_, _must not_, _required_, _shall_, _shall not_,
+_should_, _should not_, _recommended_, _may_, and _optional_ in this
+document are to be interpreted as described in [RFC
+2119](https://datatracker.ietf.org/doc/html/rfc2119).
+
+## TAP14 FORMAT
+
+TAP14's general grammar is:
+
+```ebnf
+TAPDocument := Version Plan Body | Version Body Plan
+Version     := "TAP version 14\n"
+Plan        := (Number) ".." (Number) (" # " Reason)? "\n"
```

Comment:
Is the first number ever not 1?

Reply:
There are examples I've come across in the wild of people trying to use http://npm.im/tap-parser on stuff like:

```tap
TAP version 13
4..6,10..12
ok 4
ok 6
ok 5
ok 10
ok 12
ok 11
```

I compromised and said, ok, _one_ range which can start and end wherever, and node-tap itself will only ever _emit_ a plan starting with 1.  The tap 13 specification doesn't say it _has_ to start with 1, but at the time I pushed back on allowing multiple ranges.  I don't know what they were using to emit their test data, I think just a custom thing.

If we wanna say that TAP14 requires plans be `1..N`, then I'm definitely on board with that!

Reply:
Definitely `6..4` shouldn't be allowed, so `1..0` is the only case where the "end" is lower than the "start".

Reply:
tap4j (@kinow) and node-tap (@isaacs) say plans _must_ start with 1.  Can I get a :+1: from perl (@exodist) or python (@mblayman) on this?  (Or if you think we should only go as far as "producers _should_ start plans with 1"/"harnesses _may_ reject/ignore plans if they don't start with 1", I'm fine with that, as well.)

Reply:
2fd8aaf895e58210aa173faf0826ba40bbc6cb4d

Reply:
Both Perl harnesses require it to start with a 1